### PR TITLE
Used SerDe for serializing mapjoin values and fixed NPE in selecting from partitions w/ no data

### PR DESCRIPTION
**Pull request by harveyfeng**
Fix for issues #6 and #10. 

Tested using Hive-generated mapjoin tests from src/test/tests_pass.txt (join25-40, cp_mj_rc, bucketmapjoin*, mapjoin_mapjoin, mapjoin_subquery, mapjoin_mapjoin1)

### CODE REVIEW

File: src/main/scala/shark/exec/SparkTask.scala [hunk 1]
```diff
@@ -74,8 +75,11 @@ with java.io.Serializable with LogHelper {
           work.pctx.getPrunedPartitions())
         op.parts = ppl.getConfirmedPartns.toArray ++ ppl.getUnknownPartns.toArray
         val allParts = op.parts ++ ppl.getDeniedPartns.toArray
-        op.firstConfPartDesc = Utilities.getPartitionDesc(
-          allParts(0).asInstanceOf[Partition])
+        if (allParts.size == 0) {
+	    op.firstConfPartDesc = new PartitionDesc(op.tableDesc, null)
```
Comment (rxin) [inline src/main/scala/shark/exec/SparkTask.scala]: Replace tab with space here. Same for the following line (line 81).

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 2]
```diff
@@ -118,8 +140,9 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc, HiveMapJoinOperato
     }
   }
 
-  def computeJoinKeyValuesOnPartition[T](iter: Iterator[T], posByte: Byte)
-  : Iterator[(AbstractMapJoinKey, Array[java.lang.Object])] = {
+def computeJoinKeyValuesOnPartition[T](iter: Iterator[T], posByte: Byte)
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: Indent.

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 3]
```diff
@@ -16,6 +27,7 @@ import shark.SharkEnv
 import shark.collections.Conversions._
 import spark.RDD
 import spark.broadcast.Broadcast
+import scala.collection.mutable.ArrayBuffer
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: Doesn't look like this import is necessary. If needed, add it to the scala section.

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 4]
```diff
@@ -8,6 +8,17 @@ import org.apache.hadoop.hive.ql.exec.persistence.AbstractMapJoinKey
...
+import org.apache.hadoop.hive.ql.plan.{PartitionDesc, TableDesc}
+import org.apache.hadoop.hive.ql.exec.HashTableSinkOperator.{HashTableSinkObjectCtx => MapJoinObjectCtx}
+import org.apache.hadoop.hive.ql.exec.MapJoinMetaData
+import org.apache.hadoop.hive.ql.exec.persistence.MapJoinObjectValue
+import org.apache.hadoop.hive.ql.exec.persistence.MapJoinRowContainer
+import org.apache.hadoop.hive.serde2.SerDe
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils.ObjectInspectorCopyOption
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: Get rid of the trailing semicolon.

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 5]
```diff
@@ -8,6 +8,17 @@ import org.apache.hadoop.hive.ql.exec.persistence.AbstractMapJoinKey
...
+import org.apache.hadoop.conf.Configuration
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: This is not needed.

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 6]
```diff
@@ -132,10 +155,50 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc, HiveMapJoinOperato
...
+      rowContainer.add(value)      
+      val objValue: MapJoinObjectValue = new MapJoinObjectValue(posByte,
+    		  													rowWrapper)
+      (key, objValue)
     }
   }
 
+  def setKeyMetaData() {
+    val keyTableDesc: TableDesc = conf.getKeyTblDesc()
+    val keySerializer: SerDe = keyTableDesc.getDeserializerClass().newInstance().asInstanceOf[SerDe]
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: You can get rid of the type definition here, i.e. remove ": SerDe".

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 7]
```diff
@@ -132,10 +155,50 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc, HiveMapJoinOperato
...
+    
+    keySerializer.initialize(null, keyTableDesc.getProperties())
+    
+    MapJoinMetaData.clear()
+    MapJoinMetaData.put(Integer.valueOf(metadataKeyTag), 
+        new MapJoinObjectCtx(
+        	ObjectInspectorUtils.getStandardObjectInspector(keySerializer.getObjectInspector(),
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: Don't use tabs for indenting.

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 8]
```diff
@@ -132,10 +155,50 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc, HiveMapJoinOperato
...
   }
 
+  def setKeyMetaData() {
+    val keyTableDesc: TableDesc = conf.getKeyTblDesc()
+    val keySerializer: SerDe = keyTableDesc.getDeserializerClass().newInstance().asInstanceOf[SerDe]
+    
+    keySerializer.initialize(null, keyTableDesc.getProperties())
+    
+    MapJoinMetaData.clear()
+    MapJoinMetaData.put(Integer.valueOf(metadataKeyTag),
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: A question - is metadataKeyTag always -1?

Comment (harveyfeng) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: It is - it seems like there's only one subclass instance of AbstractMapJoinKey to wrap a set of join keys, MapJoin(Single/Double)Key for 1/2 keys, and MapJoinObjectKey for >2 keys.

File: src/main/scala/shark/exec/SparkTask.scala [hunk 9]
```diff
@@ -74,8 +75,11 @@ with java.io.Serializable with LogHelper {
           work.pctx.getPrunedPartitions())
         op.parts = ppl.getConfirmedPartns.toArray ++ ppl.getUnknownPartns.toArray
         val allParts = op.parts ++ ppl.getDeniedPartns.toArray
-        op.firstConfPartDesc = Utilities.getPartitionDesc(
-          allParts(0).asInstanceOf[Partition])
+        if (allParts.size == 0) {
+	  op.firstConfPartDesc = new PartitionDesc(op.tableDesc, null)
```
Comment (rxin) [inline src/main/scala/shark/exec/SparkTask.scala]: The indent seems off after you replaced spaces with tabs? Same for line 81

File: src/main/scala/shark/exec/MapJoinOperator.scala [hunk 10]
```diff
@@ -131,11 +149,47 @@ class MapJoinOperator extends CommonJoinOperator[MapJoinDesc, HiveMapJoinOperato
         joinValuesObjectInspectors(posByte),
         joinFilters(posByte),
         joinFilterObjectInspectors(posByte),
-        noOuterJoin)
-      (key, value)
+        noOuterJoin) 
+      setValueMetaData(posByte)
+      val rowContainer = new MapJoinRowContainer[Array[Object]]()
+      rowContainer.add(value)      
+      val objValue: MapJoinObjectValue = new MapJoinObjectValue(posByte,rowContainer)
```
Comment (rxin) [inline src/main/scala/shark/exec/MapJoinOperator.scala]: nitpick, add a space after comma, before "rowContainer"

**Comment by rxin** [comment]
Looks good. Almost there :)

File: src/main/scala/shark/exec/SparkTask.scala (hunk 9 above)
Comment (harveyfeng) [inline src/main/scala/shark/exec/SparkTask.scala]: Fixed

**Pull request merged**